sycl/include/CL/sycl/properties/reduction_properties.hpp

```diff
@@ -0,0 +1,23 @@
+//==------- reduction_properties.hpp --- SYCL reduction properties ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/property_helper.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
```

Comment:
I think we should move away from using `cl::sycl` namespace. 
Can we put new API to `sycl` namespace instead of `cl::sycl`?

Reply:
> Can we put new API to sycl namespace instead of cl::sycl?

We can, but I suggest doing that in a separate patch being focusing on that re-structuring changes.
This patch is already pretty big, and already steps onto some existing/hidden rakes (see the fails caused by adding new property). I'd prefer not adding more that those.

This request also creates many various questions and comments such as:
-  the files live in include/CL/sycl directory, which is supposed to reflect the namespace hierarchy. Do you suggest creating 'include/sycl' folder and move files to there?
- Some of SYCL-2020 functionality is implemented already in cl::sycl namespace
- moving reduction out of ::cl::sycl to ::sycl might create additional effects caused by different namespaces and quick/not-well-thought solutions for it.

Reply:
Doing it in a separate patch might be considered to be a breaking API change, so I think we should resolve all these questions ASAP.

> * the files live in include/CL/sycl directory, which is supposed to reflect the namespace hierarchy. Do you suggest moving 'include/CL/sycl/reduction/hpp' to 'include/sycl/reduction.hpp'?

No, I don't suggest that. Moreover, I think it's DPC++ implementation detail - SYCL specification defines only `sycl/sycl.hpp` header.
I see that @romanovvlad already added sycl/sycl.hpp in https://github.com/intel/llvm/commit/5edb228a9e60260bbc158030cfd7c1a0ef4b1e1d. Vlad what is the long term plan for migration to SYCL-2020 API?

> * Some of SYCL-2020 functionality is implemented already in cl::sycl namespace

Okay, but how it's relevant to my request?

> * moving reduction out of :🆑:sycl to ::sycl might create additional effects caused by different namespaces and quick/not-well-thought solutions for it.

Could you clarify what effects do you mean, please?

Reply:
> Vlad what is the long term plan for migration to SYCL-2020 API?

There is no long term plan. We need to support SYCL1.2.1 so far.

Reply:
> Could you clarify what effects do you mean, please?

When I wrote that statement I was not sure about what effects. Now after trying the new reduction in ::sycl namespace in my local workspace I see many surprising LIT fails like this:

FAIL: SYCL :: abi/layout_array.cpp (11 of 204)
/ws/gs_public/llvm/sycl/test/abi/layout_array.cpp:34:11: error: CHECK: expected string not found in input
// CHECK: 0 | class sycl::item<2, true>

FAIL: SYCL :: abi/vtable.cpp (12 of 204)
/ws/gs_public/llvm/sycl/test/abi/vtable.cpp:17:11: error: CHECK: expected string not found in input
// CHECK: Vtable for 'sycl::detail::HostKernelBase' (6 entries).

FAIL: SYCL :: basic_tests/vectors/ctad.cpp (59 of 204)
  File /ws/gs_public/llvm/sycl/test/basic_tests/vectors/ctad.cpp Line 12: redefinition of 'sycl'

FAIL: SYCL :: multi_ptr/ctad.cpp (105 of 204)
  File /ws/gs_public/llvm/sycl/test/multi_ptr/ctad.cpp Line 12: redefinition of 'sycl'

FAIL: SYCL :: regression/builtins_vector1.cpp (114 of 204)
/ws/gs_public/llvm/sycl/test/regression/builtins_vector1.cpp:6:11: error: redefinition of 'sycl'
namespace sycl = cl::sycl;

and so on.

I can fix LIT tests, but similar things may appear in user's apps as well

Reply:
>> Some of SYCL-2020 functionality is implemented already in cl::sycl namespace

> Okay, but how it's relevant to my request?

You can use older buffer class by calling "sycl::buffer" because the namespace cl is inline namespace.
I doubt SYCL-2020 class buffer would ever be duplicated in true ::sycl namespace.
Some other mechanism for sycl-2020 buffer will be used (probably 'inline namespace cl' is already sufficient for that).
So, why use different approach for the new SYCL2020 reduction class? 
Using one approach for 'buffer' (and other classes) and a different one for 'reduction' would only give additional risks.

Reply:
> If you use it, all SYCL types, etc. are available in the `cl::sycl` namespace.

Does it include new types like kernel_bundle?

Reply:
I do not think the spirit of the SYCL 2020 specification is to add new features inside the deprecated `::cl::sycl`.
But I can admit that it is not forbidden just for developer convenience.
As deprecated, I guess we cannot check for anything in `::cl::sycl` in the CTS...